[tests] Split NodeConn from NodeConnCB

jnewbery opened the pull request:
This is the final step in #11518, except for possibly renaming - for motivation, please see that PR.

If this is merged, then migrating the test framework from asyncore to asyncio should be easier (I say should because I haven't dug too deeply into what would be required).

Requesting review from @ryanofsky , since he always has good feedback on these refactor PRs, and I'd appreciate his take on this refactor. Note particularly that I've reverted the change suggested here: https://github.com/bitcoin/bitcoin/pull/11182#discussion_r148859555 . The idea, as always, is to present a simple interface to the test writer.

jnewbery committed:
[tests] Remove rpc property from TestNode in p2p-segwit.py.

Change the helper methods to functions which take a node and a p2p
connection as arguments.

jnewbery committed:
[tests] Move version message sending from NodeConn to NodeConnCB

This commit moves the logic that sends a version message
on connection from NodeConn to NodeConnCB. NodeConn should
not be aware of the semantics or meaning of the P2P payloads.

jnewbery committed:
[tests] Make NodeConnCB a subclass of NodeConn

This makes NodeConnCB a subclass of NodeConn, and
removes the need for the client code to know
anything about the implementation details of NodeConnCB.

NodeConn can now be swapped out for any other implementation
of a low-level connection without changing client code.

ryanofsky commented on test/functional/example_test.py:
```diff
@@ -34,7 +35,7 @@
 # override the on_*() methods if you need custom behaviour.
 class BaseNode(NodeConnCB):
-    def __init__(self):
+    def __init__(self, dstaddr, dstport, net="regtest", services=NODE_NETWORK, send_version=True):
```
In commit "[tests] Make NodeConnCB a subclass of NodeConn"

Can get rid of all these default parameters repeated so many times. See main comment.

ryanofsky reviewed (commented):
This basically looks good, but I really think it's important to stop creating peer objects via callback in add_p2p_connection. The change I think you should make is here a683588fea93681f8dae67b1532532ee7b19884a (previously described in https://github.com/bitcoin/bitcoin/pull/11182#discussion_r148859555).

Advantages of this change:

- Simpler control flow. Tests create their own peer objects. You can see when and where they get constructed and what arguments they are constructed with. No need to pass around callbacks. There's no advantage I can see to passing around callbacks unless saving two `()` characters is an advantage.

- Less coupling and repetition. TestNode no longer needs need to know anything about peer objects other than that they have connect and disconnect methods. The connect arguments (dstaddr, dstport, net="regtest", services=NODE_NETWORK, send_version=True) are defined and used one place instead of repeated 12 times in testnode, mininode, and comptool code, and in individual test case code.

- Lets you drop second commit, reduces overall diff from 502 lines to 457 lines.

jnewbery committed:
[tests] Remove mininode periodic (half-hour) ping messages

jnewbery committed:
[tests] Tidy up mininode

Add docstrings and renames some methods.
Also removes the redundant NodeConn.readable() method override.

jnewbery committed:
[tests] Move only: move NodeConnCB below NodeConn

This is required since NodeConnCB will inherit from NodeConn
after the next commit.

jnewbery commented:
Thanks Russ. I much prefer your suggested way of passing the constructed NodeConnCB object to TestNode. I've incorporated your change and split the PR into smaller commits to aid reviewing.

MarcoFalke commented on test/functional/test_framework/test_node.py:
```diff
@@ -185,10 +184,8 @@ def p2p(self):
-                p.connection.disconnect_node()
-        self.p2ps = []
+            p.peer_disconnect()
+        del self.p2ps[:]
```
stylte-nit: Any reason for this change? I think the previous is easier to read python code.

jnewbery replied:
There's some discussion of the differences here: https://stackoverflow.com/questions/850795/different-ways-of-clearing-lists

If there was a different reference to this array, then `self.p2ps = []` wouldn't clear that array, it would just change `self.p2ps` to point to a new, empty array.

Practically, there's not much difference, but I think I prefer `del self.p2ps[:]`. I actually got this from @ryanofsky's commit - perhaps he had a specific reason to make this change?

MarcoFalke replied:
Ok, fine to keep, since it is ever-so-slightly safer and already used by this syntax in `stop_node`

ryanofsky replied:
Main reason for suggesting this is that having multiple instances of the same list can make things confusing when you are debugging.

maflcko reviewed (commented):
Concept ACK a328ab35888baaa481b5b59e41472f91e8802f8e. Going to review this week or so.

ryanofsky commented on test/functional/test_framework/mininode.py:
```diff
@@ -241,15 +258,19 @@ def _log_message(self, direction, msg):
+    def peer_connect(self, *args, **kwargs):
+        super().peer_connect(*args, **kwargs)
 
         # Track number of messages of each type received and the most recent
```
In commit "[tests] Make NodeConnCB a subclass of NodeConn"

Maybe keep assigning nServices/messages_count/last_message/ping_counter members that don't depend on connection information in `__init__` rather than `peer_connect`. This would mean you don't need to introduce a `NodeConnCB.peer_connect` method in this commit, and in the next commit it can be a shorter, more focused override. Just a suggestion, though, difference is minor.

jnewbery replied:
Agree. Makes more sense. Thanks.

ryanofsky reviewed (commented):
utACK a328ab35888baaa481b5b59e41472f91e8802f8e

jnewbery commented on test/functional/test_framework/mininode.py:
```diff
@@ -262,12 +264,18 @@ def readable(self):
         return True
 
     def handle_read(self):
+        """asyncore callback when data is read from the socket."""
```
We can actually just remove the `readable()` override entirely. From https://docs.python.org/3.6/library/asyncore.html#asyncore.dispatcher.readable : "The default method simply returns True, indicating that by default, all channels will be interested in read events." so this override isn't actually doing anything.

I'll update the commit.

jnewbery commented on test/functional/test_framework/mininode.py:
```diff
@@ -262,12 +264,18 @@ def readable(self):
+            self.read_message()
 
-    def got_data(self):
+    def read_message(self):
```
Agree. Changed to `_on_data`.

jnewbery commented on test/functional/test_framework/mininode.py:
```diff
@@ -262,12 +264,18 @@ def readable(self):
 
-    def got_data(self):
+    def read_message(self):
+        """Try to read a P2P message from the recv buffer.
```
Yes. Fixed.

jnewbery commented:
Addressed @ryanofsky's comments

ryanofsky reviewed (commented):
utACK e9dfa9bccc5cbb6096c60498651b451297f0a931. Only changes since last review were the various things commented above.

MarcoFalke commented on test/functional/test_framework/test_node.py:
```diff
@@ -158,7 +157,7 @@ def node_encrypt_wallet(self, passphrase):
         self.wait_until_stopped()
 
-    def add_p2p_connection(self, p2p_conn, **kwargs):
+    def add_p2p_connection(self, p2p_conn, *args, **kwargs):
```
nit: I'd rather do the opposite and disallow passing through unnamed args here. Forcing code to be explicit should be safer and ease future review, imo.

 ```py
def add_p2p_connection(self, p2p_conn, *, **kwargs)
```

ryanofsky replied:
If the goal is to disallow unnamed arguments, this place to do is where the arguments are actually declared (in the peer_connect method). If you make that change here instead of there, you will be introducing an inconsistency between add_p2p_connection and peer_connect. It is better for testnode to just forward arguments on to peernode, instead making unnecessary assumptions about how it works internally and creating unnecessary differences. Forwarding *args and **kwargs is also idiomatic in python while forwarding just **kwargs is more unusual.

jnewbery replied:
I think @ryanofsky is right - the right place for this is in `NodeConnCB.peer_connect()`. Can we leave that for a future PR?

MarcoFalke replied:
@jnewbery Sure, future PR sounds good.

maflcko reviewed (commented):
utACK e9dfa9bccc5cbb6096c60498651b451297f0a931
